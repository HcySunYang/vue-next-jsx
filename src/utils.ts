import * as bt from '@babel/types'

export function makeMap(
  str: string,
  expectsLowerCase?: boolean
): (key: string) => boolean {
  const map: Record<string, boolean> = Object.create(null)
  const list: Array<string> = str.split(',')
  for (let i = 0; i < list.length; i++) {
    map[list[i]] = true
  }
  return expectsLowerCase
    ? (val) => !!map[val.toLowerCase()]
    : (val) => !!map[val]
}

// https://developer.mozilla.org/en-US/docs/Web/HTML/Element
const HTML_TAGS =
  'html,body,base,head,link,meta,style,title,address,article,aside,footer,' +
  'header,h1,h2,h3,h4,h5,h6,hgroup,nav,section,div,dd,dl,dt,figcaption,' +
  'figure,picture,hr,img,li,main,ol,p,pre,ul,a,b,abbr,bdi,bdo,br,cite,code,' +
  'data,dfn,em,i,kbd,mark,q,rp,rt,rtc,ruby,s,samp,small,span,strong,sub,sup,' +
  'time,u,var,wbr,area,audio,map,track,video,embed,object,param,source,' +
  'canvas,script,noscript,del,ins,caption,col,colgroup,table,thead,tbody,td,' +
  'th,tr,button,datalist,fieldset,form,input,label,legend,meter,optgroup,' +
  'option,output,progress,select,textarea,details,dialog,menu,menuitem,' +
  'summary,content,element,shadow,template,blockquote,iframe,tfoot'

// https://developer.mozilla.org/en-US/docs/Web/SVG/Element
const SVG_TAGS =
  'svg,animate,animateMotion,animateTransform,circle,clipPath,color-profile,' +
  'defs,desc,discard,ellipse,feBlend,feColorMatrix,feComponentTransfer,' +
  'feComposite,feConvolveMatrix,feDiffuseLighting,feDisplacementMap,' +
  'feDistanceLight,feDropShadow,feFlood,feFuncA,feFuncB,feFuncG,feFuncR,' +
  'feGaussianBlur,feImage,feMerge,feMergeNode,feMorphology,feOffset,' +
  'fePointLight,feSpecularLighting,feSpotLight,feTile,feTurbulence,filter,' +
  'foreignObject,g,hatch,hatchpath,image,line,linearGradient,marker,mask,' +
  'mesh,meshgradient,meshpatch,meshrow,metadata,mpath,path,pattern,' +
  'polygon,polyline,radialGradient,rect,set,solidcolor,stop,switch,symbol,' +
  'text,textPath,title,tspan,unknown,use,view'

export const isHTMLTag = makeMap(HTML_TAGS)
export const isSVGTag = makeMap(SVG_TAGS)

// Transform `JSXMemberExpression` to `MemberExpression`
export function transformJSXMemberExpression(
  jsxMemberExpress: bt.JSXMemberExpression
): bt.MemberExpression {
  return bt.memberExpression(
    bt.isJSXIdentifier(jsxMemberExpress.object)
      ? bt.identifier(jsxMemberExpress.object.name)
      : transformJSXMemberExpression(jsxMemberExpress.object),
    bt.identifier(jsxMemberExpress.property.name)
  )
}

const cacheStringFunction = <T extends (str: string) => string>(fn: T): T => {
  const cache: Record<string, string> = Object.create(null)
  return ((str: string) => {
    const hit = cache[str]
    return hit || (cache[str] = fn(str))
  }) as any
}

const camelizeRE = /-(\w)/g
export const camelize = cacheStringFunction((str: string): string => {
  return str.replace(camelizeRE, (_, c) => (c ? c.toUpperCase() : ''))
})

export const capitalize = cacheStringFunction((str: string): string => {
  return str.charAt(0).toUpperCase() + str.slice(1)
})
