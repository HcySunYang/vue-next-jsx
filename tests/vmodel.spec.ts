import { transformWithPlugin } from './utils'

describe('v-model', () => {
  describe('Component: ', () => {
    test('default modelValue', () => {
      const code = `<Comp v-model={ ref.val } />`
      transformWithPlugin(code)
    })

    test('default modelValue (with modifiers)', () => {
      const code = `<Comp v-model_a_b={ ref.val } />`
      transformWithPlugin(code)
    })

    test('specify model prop name', () => {
      const code = `<Comp v-model-foo={ ref.val } />`
      transformWithPlugin(code)
    })

    test('specify model prop name (with modifiers)', () => {
      const code = `<Comp v-model-foo_a_b={ ref.val } />`
      transformWithPlugin(code)
    })

    test('specify model prop name (with other props)', () => {
      const code = `<Comp v-model-foo_a_b={ ref.val } a="b" />`
      transformWithPlugin(code)
    })

    test('specify model prop name (with v-on)', () => {
      const code = `<Comp v-model-foo_a_b={ ref.val } v-on={ obj } />`
      transformWithPlugin(code)
    })
  })

  describe('Element: ', () => {
    test('input with v-model', () => {
      const code = `<input v-model={ ref.val } />`
      transformWithPlugin(code)
    })

    test('textarea with v-model', () => {
      const code = `<textarea v-model={ ref.val } />`
      transformWithPlugin(code)
    })

    test('input with dynamic type attribute', () => {
      const code = `<input v-model={ ref.val } type={ refType.value } />`
      transformWithPlugin(code)
    })

    test('input with type=radio', () => {
      const code = `<input v-model={ ref.val } type="radio" />`
      transformWithPlugin(code)
    })

    test('input with type=checkbox', () => {
      const code = `<input v-model={ ref.val } type="checkbox" />`
      transformWithPlugin(code)
    })

    test('input with dynamic Key binding', () => {
      const code = `<input v-model={ ref.val } { ...props } />`
      transformWithPlugin(code)
    })
  })
})
