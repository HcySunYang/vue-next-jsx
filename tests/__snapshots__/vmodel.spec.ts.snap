// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v-model Component:  default modelValue (with modifiers) 1`] = `
"createVNode(Comp, {
  \\"modelValue\\": ref.val,
  \\"onUpdate:modelValue\\": $event => ref.val = $event,
  \\"modelModifiers\\": {
    \\"a\\": true,
    \\"b\\": true
  }
}, null);"
`;

exports[`v-model Component:  default modelValue 1`] = `
"createVNode(Comp, {
  \\"modelValue\\": ref.val,
  \\"onUpdate:modelValue\\": $event => ref.val = $event
}, null);"
`;

exports[`v-model Component:  specify model prop name (with modifiers) 1`] = `
"createVNode(Comp, {
  \\"foo\\": ref.val,
  \\"onUpdate:foo\\": $event => ref.val = $event,
  \\"fooModifiers\\": {
    \\"a\\": true,
    \\"b\\": true
  }
}, null);"
`;

exports[`v-model Component:  specify model prop name (with other props) 1`] = `
"createVNode(Comp, {
  \\"foo\\": ref.val,
  \\"onUpdate:foo\\": $event => ref.val = $event,
  \\"fooModifiers\\": {
    \\"a\\": true,
    \\"b\\": true
  },
  \\"a\\": \\"b\\"
}, null);"
`;

exports[`v-model Component:  specify model prop name (with v-on) 1`] = `
"createVNode(Comp, mergeProps({
  \\"foo\\": ref.val,
  \\"onUpdate:foo\\": $event => ref.val = $event,
  \\"fooModifiers\\": {
    \\"a\\": true,
    \\"b\\": true
  }
}, toHandlers(obj)), null);"
`;

exports[`v-model Component:  specify model prop name 1`] = `
"createVNode(Comp, {
  \\"foo\\": ref.val,
  \\"onUpdate:foo\\": $event => ref.val = $event
}, null);"
`;

exports[`v-model Element:  input with dynamic Key binding 1`] = `
"withDirectives(createVNode(\\"input\\", mergeProps({
  \\"onUpdate:modelValue\\": $event => ref.val = $event
}, { ...props
}), null), [vModelDynamic, ref.val]);"
`;

exports[`v-model Element:  input with dynamic type attribute 1`] = `
"withDirectives(createVNode(\\"input\\", {
  \\"onUpdate:modelValue\\": $event => ref.val = $event,
  \\"type\\": refType.value
}, null), [vModelDynamic, ref.val]);"
`;

exports[`v-model Element:  input with type=checkbox 1`] = `
"withDirectives(createVNode(\\"input\\", {
  \\"onUpdate:modelValue\\": $event => ref.val = $event,
  \\"type\\": \\"checkbox\\"
}, null), [vModelCheckbox, ref.val]);"
`;

exports[`v-model Element:  input with type=radio 1`] = `
"withDirectives(createVNode(\\"input\\", {
  \\"onUpdate:modelValue\\": $event => ref.val = $event,
  \\"type\\": \\"radio\\"
}, null), [vModelRadio, ref.val]);"
`;

exports[`v-model Element:  input with v-model 1`] = `
"withDirectives(createVNode(\\"input\\", {
  \\"onUpdate:modelValue\\": $event => ref.val = $event
}, null), [vModelText, ref.val]);"
`;

exports[`v-model Element:  textarea with v-model 1`] = `
"withDirectives(createVNode(\\"textarea\\", {
  \\"onUpdate:modelValue\\": $event => ref.val = $event
}, null), [vModelText, ref.val]);"
`;
